### DISCUSSION TITLE
How to use repository in entity for validation or other situation?

### DISCUSSION THREAD

**Discussion by CAH-FlyChen** [abpframework/abp#10472] [Q&A] [answered]
I want to return product name in Lock methord of stock entity. How should deal with the  _productRepository ??
1.use "stockValidatorDomainService" pass into the Lock methord as param?
2.use specification?  but it dos not have async methord in specification<T>?

````csharp

public class stock:entity
{
       public async Task Lock(Zion.ERP.Domain.Stock.Stock stock, int qty, string batchNo)
        {
            var item = stock.Details.Single(t => t.BatchNo == batchNo);
            if (qty > stock.AvalableQty)
            {
                var prod = (await _productRepository.WithDetailsAsync(t => t.ProductTemplate)).Single(t => t.Id == stock.ProductId);
                       throw new ZionCommonException($"product{prod.ProductTemplate.Name} [{prod.ProductTemplate.MainOE}],batch:{batchNo} is not enough");
            }
            stock.LockedQty += qty; 
            item.LockQty(qty);
        }
}
```

**Comment by EngincanV** [comment, marked as answer]
Hi @CAH-FlyChen, there are two things you can do,

* Firstly, you can pass the `IProductRepository` interface as a parameter to the **Lock** method 

```csharp
public class stock:entity
{
       public async Task Lock(IProductRepository productRepository, Stock stock, int qty, string batchNo)
        {
            var item = stock.Details.Single(t => t.BatchNo == batchNo);
            if (qty > stock.AvalableQty)
            {
                var prod = (await productRepository.WithDetailsAsync(t => t.ProductTemplate)).Single(t => t.Id == stock.ProductId);
                       throw new ZionCommonException($"product{prod.ProductTemplate.Name} [{prod.ProductTemplate.MainOE}],batch:{batchNo} is not enough");
            }
            stock.LockedQty += qty; 
            item.LockQty(qty);
        }
}
```

* or secondly, you can create a [`Domain Service`](https://docs.abp.io/en/abp/5.0/Domain-Services), manage the validity of your entities from there and call the method that you've defined in the **Domain Service** on your Application Service (where it is needed).

   * Remove the lock method from your entity method (it's good to make your entity class clean) and you can change the related properties' setter as `internal set` like in below.
```csharp
public class stock:entity
{
       //remove the lock method and make related properties' setter as `internal set`
       public int LockedQty { get; internal set; } //internal set -> ensure only can be set in Domain Layer
}
```

   * Create a **Domain Service** and move your **Lock** method in this class. Then, you can inject one or more services into this class to self-validate your entity. For more info please read the [Domain Service documentation](https://docs.abp.io/en/abp/5.0/Domain-Services).
```csharp
public class StockManager: DomainService
{
    private readonly IProductRepository _productRepository;

    public IssueManager(IProductRepository productRepository)
    {
        _productRepository = productRepository;
    }

   public async Task Lock(Stock stock, int qty, string batchNo)
        {
            var item = stock.Details.Single(t => t.BatchNo == batchNo);
            if (qty > stock.AvalableQty)
            {
                var prod = (await _productRepository.WithDetailsAsync(t => t.ProductTemplate)).Single(t => t.Id == stock.ProductId);
                       throw new ZionCommonException($"product{prod.ProductTemplate.Name} [{prod.ProductTemplate.MainOE}],batch:{batchNo} is not enough");
            }
            stock.LockedQty += qty; 
            item.LockQty(qty);
        }
}    
```

--- 
> 2.use specification? but it dos not have async methord in specification?

[Specification Pattern](https://docs.abp.io/en/abp/5.0/Specifications) is used to define named, reusable, combinable and testable **filters** for entities and other business objects.

* As the description pointed out, it's only used to create a common filter that can be used more than one time to filter. Therefore, it does not need any async method in its implementation, because it's not run any code under the hook. It basically helps to define common filter expressions.

* You can read the [Specification](https://docs.abp.io/en/abp/5.0/Specifications) document to see its usage.

**Reply by CAH-FlyChen** [reply]
Got it ! Thanks  @EngincanV
